ConsulTxnStore: panic when get transaction returns error

After more real-world testing of [the new transaction-based Consul KV store](https://github.com/openark/orchestrator/pull/1276) I encountered a panic when the get/read transaction fails with an error:
```
2021-02-04 05:06:27 DEBUG kv.SubmitMastersToKvStores, clusterName: , force: false: numPairs: 525
2021-02-04 05:06:27 DEBUG kv.SubmitMastersToKvStores: submitKvPairs: 0
2021-02-04 05:06:27 DEBUG consulTxnStore.DistributePairs(): distributing 525 pairs to 20 datacenters
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x921e22]

goroutine 92749 [running]:
github.com/openark/orchestrator/go/kv.(*consulTxnStore).updateDatacenterKVPairs(0xc00066e060, 0xc00744d280, 0xc00744c720, 0x7, 0xc00ac0e000, 0x20d, 0x400, 0x5, 0x0, 0x0, ...)
	/tmp/brew2deb/packages/orchestrator/src/orchestrator-gh.git/.gopath/src/github.com/openark/orchestrator/go/kv/consul_txn.go:120 +0xa02
github.com/openark/orchestrator/go/kv.(*consulTxnStore).DistributePairs(0xc00066e060, 0xc008ace000, 0x20d, 0x300, 0x0, 0x0)
	/tmp/brew2deb/packages/orchestrator/src/orchestrator-gh.git/.gopath/src/github.com/openark/orchestrator/go/kv/consul_txn.go:229 +0x498
github.com/openark/orchestrator/go/kv.DistributePairs(0xc008ace000, 0x20d, 0x300, 0xc00a499f48, 0x1)
	/tmp/brew2deb/packages/orchestrator/src/orchestrator-gh.git/.gopath/src/github.com/openark/orchestrator/go/kv/kv.go:109 +0x85
github.com/openark/orchestrator/go/logic.SubmitMastersToKvStores(0x0, 0x0, 0xc000584500, 0xc007d5b76d, 0x449a96, 0xcde850, 0xc007d5b7d0, 0x9ddba5, 0xc0001ea6b0)
	/tmp/brew2deb/packages/orchestrator/src/orchestrator-gh.git/.gopath/src/github.com/openark/orchestrator/go/logic/orchestrator.go:462 +0x59f
created by github.com/openark/orchestrator/go/logic.ContinuousDiscovery.func5
	/tmp/brew2deb/packages/orchestrator/src/orchestrator-gh.git/.gopath/src/github.com/openark/orchestrator/go/logic/orchestrator.go:582 +0x316
```

This happens because [go/kv/consul_txn.go:120](https://github.com/openark/orchestrator/blob/master/go/kv/consul_txn.go#L120) is not checking if `getTxnResp` is nil, which happens if the get transaction fails

This PR updates the logic to consider this failure and adds better unit testing of the ConsulTxnStore. Following this change if there is a failure to read/get the existing keys the ConsulTxnStore will still attempt send updates to the KVs without panic'ing

A PR to fix this coming soon!

--- a/go/kv/consul_txn.go
+++ b/go/kv/consul_txn.go
@@ -71,13 +71,12 @@ func NewConsulTxnStore() KVStore {
 func (this *consulTxnStore) doWriteTxn(txnOps consulapi.TxnOps, queryOptions *consulapi.QueryOptions) (err error) {
 	ok, resp, _, err := this.client.Txn().Txn(txnOps, queryOptions)
 	if err != nil {
+		log.Errorf("consulTxnStore.doWriteTxn(): failed %v", err)
 		return err
 	} else if !ok {
-		// return the first transaction error found
-		for _, txnErr := range resp.Errors {
-			if txnErr.What != "" {
-				return fmt.Errorf("consul txn error: %v", txnErr.What)
-			}
+		for _, terr := range resp.Errors {
+			log.Errorf("consulTxnStore.doWriteTxn(): transaction error %v", terr.What)
+			err = fmt.Errorf("%v", terr.What)
 		}
 	}
 	return err

--- a/go/kv/consul_txn.go
+++ b/go/kv/consul_txn.go
@@ -90,7 +90,9 @@ func (this *consulTxnStore) updateDatacenterKVPairs(wg *sync.WaitGroup, dc strin
 	kcCacheKeys := make([]string, 0)
 
 	// get the current key-values in a single transaction
+	var terr error
 	var getTxnOps consulapi.TxnOps
+	var getTxnResp *consulapi.TxnResponse
 	var possibleSetKVPairs []*consulapi.KVPair
 	for _, kvPair := range kvPairs {
 		val := string(kvPair.Value)
@@ -108,21 +110,25 @@ func (this *consulTxnStore) updateDatacenterKVPairs(wg *sync.WaitGroup, dc strin
 		})
 		possibleSetKVPairs = append(possibleSetKVPairs, kvPair)
 	}
-	_, getTxnResp, _, e := this.client.Txn().Txn(getTxnOps, queryOptions)
-	if err != nil {
-		err = e
+	if len(getTxnOps) > 0 {
+		_, getTxnResp, _, terr = this.client.Txn().Txn(getTxnOps, queryOptions)
+		if terr != nil {
+			err = terr
+		}
 	}
 
 	// find key-value pairs that need updating, add pairs that need updating to set transaction
 	var setTxnOps consulapi.TxnOps
 	for _, pair := range possibleSetKVPairs {
 		var kvExists bool
-		for _, result := range getTxnResp.Results {
-			if pair.Key == result.KV.Key && string(pair.Value) == string(result.KV.Value) {
-				existing++
-				kvExists = true
-				this.kvCache.SetDefault(getConsulKVCacheKey(dc, pair.Key), string(pair.Value))
-				break
+		if getTxnResp != nil && len(getTxnResp.Errors) == 0 {
+			for _, result := range getTxnResp.Results {
+				if pair.Key == result.KV.Key && string(pair.Value) == string(result.KV.Value) {
+					existing++
+					kvExists = true
+					this.kvCache.SetDefault(getConsulKVCacheKey(dc, pair.Key), string(pair.Value))
+					break
+				}
 			}
 		}
 		if !kvExists {
@@ -138,10 +144,10 @@ func (this *consulTxnStore) updateDatacenterKVPairs(wg *sync.WaitGroup, dc strin
 
 	// update key-value pairs in a single Consul Transaction
 	if len(setTxnOps) > 0 {
-		if e := this.doWriteTxn(setTxnOps, queryOptions); e != nil {
+		if terr = this.doWriteTxn(setTxnOps, queryOptions); terr != nil {
 			log.Errorf("consulTxnStore.DistributePairs(): failed %v", kcCacheKeys)
 			failed = len(setTxnOps)
-			err = e
+			err = terr
 		} else {
 			for _, txnOp := range setTxnOps {
 				this.kvCache.SetDefault(getConsulKVCacheKey(dc, txnOp.KV.Key), string(txnOp.KV.Value))

--- a/go/kv/consul_txn.go
+++ b/go/kv/consul_txn.go
@@ -1,5 +1,5 @@
 /*
-   Copyright 2020 Shlomi Noach, GitHub Inc.
+   Copyright 2021 Shlomi Noach, GitHub Inc.
 
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
@@ -114,7 +114,7 @@ func (this *consulTxnStore) updateDatacenterKVPairs(wg *sync.WaitGroup, dc strin
 	if len(getTxnOps) > 0 {
 		_, getTxnResp, _, terr = this.client.Txn().Txn(getTxnOps, queryOptions)
 		if terr != nil {
-			log.Errorf("consulTxnStore.DistributePairs(): get transaction failed %v", terr)
+			log.Errorf("consulTxnStore.DistributePairs(): %v", terr)
 			err = terr
 		}
 	}
@@ -124,11 +124,6 @@ func (this *consulTxnStore) updateDatacenterKVPairs(wg *sync.WaitGroup, dc strin
 	for _, pair := range possibleSetKVPairs {
 		var kvExistsAndEqual bool
 		if getTxnResp != nil {
-			for _, terr := range getTxnResp.Errors {
-				if !strings.HasSuffix(terr.What, "doesn't exist") {
-					log.Errorf("consulTxnStore.DistributePairs(): get transaction error %v", terr.What)
-				}
-			}
 			for _, result := range getTxnResp.Results {
 				if pair.Key == result.KV.Key && string(pair.Value) == string(result.KV.Value) {
 					this.kvCache.SetDefault(getConsulKVCacheKey(dc, pair.Key), string(pair.Value))

--- a/go/kv/consul_txn.go
+++ b/go/kv/consul_txn.go
@@ -20,7 +20,6 @@ import (
 	"crypto/tls"
 	"fmt"
 	"net/http"
-	"strings"
 	"sync"
 	"sync/atomic"
 

--- a/go/kv/consul_txn.go
+++ b/go/kv/consul_txn.go
@@ -114,6 +114,7 @@ func (this *consulTxnStore) updateDatacenterKVPairs(wg *sync.WaitGroup, dc strin
 	if len(getTxnOps) > 0 {
 		_, getTxnResp, _, terr = this.client.Txn().Txn(getTxnOps, queryOptions)
 		if terr != nil {
+			log.Errorf("consulTxnStore.DistributePairs(): get transaction failed %v", terr)
 			err = terr
 		}
 	}
@@ -122,13 +123,12 @@ func (this *consulTxnStore) updateDatacenterKVPairs(wg *sync.WaitGroup, dc strin
 	var setTxnOps consulapi.TxnOps
 	for _, pair := range possibleSetKVPairs {
 		var kvExistsAndEqual bool
-		if len(getTxnResp.Errors) > 0 {
+		if getTxnResp != nil {
 			for _, terr := range getTxnResp.Errors {
 				if !strings.HasSuffix(terr.What, "doesn't exist") {
 					log.Errorf("consulTxnStore.DistributePairs(): get transaction error %v", terr.What)
 				}
 			}
-		} else if getTxnResp != nil {
 			for _, result := range getTxnResp.Results {
 				if pair.Key == result.KV.Key && string(pair.Value) == string(result.KV.Value) {
 					this.kvCache.SetDefault(getConsulKVCacheKey(dc, pair.Key), string(pair.Value))

--- a/go/kv/consul_txn.go
+++ b/go/kv/consul_txn.go
@@ -20,6 +20,7 @@ import (
 	"crypto/tls"
 	"fmt"
 	"net/http"
+	"strings"
 	"sync"
 	"sync/atomic"
 
@@ -75,7 +76,8 @@ func (this *consulTxnStore) doWriteTxn(txnOps consulapi.TxnOps, queryOptions *co
 		return err
 	} else if !ok {
 		for _, terr := range resp.Errors {
-			log.Errorf("consulTxnStore.doWriteTxn(): transaction error %v", terr.What)
+			txnOp := txnOps[terr.OpIndex]
+			log.Errorf("consulTxnStore.doWriteTxn(): transaction error %q for KV %s on %s", terr.What, txnOp.KV.Verb, txnOp.KV.Key)
 			err = fmt.Errorf("%v", terr.What)
 		}
 	}
@@ -119,18 +121,24 @@ func (this *consulTxnStore) updateDatacenterKVPairs(wg *sync.WaitGroup, dc strin
 	// find key-value pairs that need updating, add pairs that need updating to set transaction
 	var setTxnOps consulapi.TxnOps
 	for _, pair := range possibleSetKVPairs {
-		var kvExists bool
-		if getTxnResp != nil && len(getTxnResp.Errors) == 0 {
+		var kvExistsAndEqual bool
+		if len(getTxnResp.Errors) > 0 {
+			for _, terr := range getTxnResp.Errors {
+				if !strings.HasSuffix(terr.What, "doesn't exist") {
+					log.Errorf("consulTxnStore.DistributePairs(): get transaction error %v", terr.What)
+				}
+			}
+		} else if getTxnResp != nil {
 			for _, result := range getTxnResp.Results {
 				if pair.Key == result.KV.Key && string(pair.Value) == string(result.KV.Value) {
-					existing++
-					kvExists = true
 					this.kvCache.SetDefault(getConsulKVCacheKey(dc, pair.Key), string(pair.Value))
+					kvExistsAndEqual = true
+					existing++
 					break
 				}
 			}
 		}
-		if !kvExists {
+		if !kvExistsAndEqual {
 			setTxnOps = append(setTxnOps, &consulapi.TxnOp{
 				KV: &consulapi.KVTxnOp{
 					Verb:  consulapi.KVSet,
@@ -143,10 +151,9 @@ func (this *consulTxnStore) updateDatacenterKVPairs(wg *sync.WaitGroup, dc strin
 
 	// update key-value pairs in a single Consul Transaction
 	if len(setTxnOps) > 0 {
-		if terr = this.doWriteTxn(setTxnOps, queryOptions); terr != nil {
+		if err = this.doWriteTxn(setTxnOps, queryOptions); err != nil {
 			log.Errorf("consulTxnStore.DistributePairs(): failed %v", kcCacheKeys)
 			failed = len(setTxnOps)
-			err = terr
 		} else {
 			for _, txnOp := range setTxnOps {
 				this.kvCache.SetDefault(getConsulKVCacheKey(dc, txnOp.KV.Key), string(txnOp.KV.Value))

--- a/go/kv/consul_txn_test.go
+++ b/go/kv/consul_txn_test.go
@@ -85,8 +85,8 @@ func TestConsulTxnStorePutKVPairs(t *testing.T) {
 		if err := store.PutKVPairs([]*KVPair{
 			{Key: "fail1", Value: "test"},
 			{Key: "fail2", Value: "test2"},
-		}); err == nil || err.Error() != "consul txn error: test error" {
-			t.Fatalf("Expected %q error from .PutKVPairs(), got: %q", "consul txn error: test error", err.Error())
+		}); err == nil || err.Error() != "test error" {
+			t.Fatalf("Expected %q error from .PutKVPairs(), got: %q", "test error", err.Error())
 		}
 	})
 }

--- a/go/kv/consul_txn_test.go
+++ b/go/kv/consul_txn_test.go
@@ -107,10 +107,10 @@ func TestConsulTxnStoreUpdateDatacenterKVPairs(t *testing.T) {
 			Response: &consulapi.TxnResponse{
 				Results: consulapi.TxnResults{
 					{
-						KV: &consulapi.KVPair{Key: "test", Value: []byte("test"), ModifyIndex: 123},
+						KV: &consulapi.KVPair{Key: "test", Value: []byte("test")},
 					},
 					{
-						KV: &consulapi.KVPair{Key: "test2", Value: []byte("not-equal"), ModifyIndex: 123},
+						KV: &consulapi.KVPair{Key: "test2", Value: []byte("not-equal")},
 					},
 				},
 			},

--- a/go/kv/consul_txn_test.go
+++ b/go/kv/consul_txn_test.go
@@ -98,19 +98,43 @@ func TestConsulTxnStoreUpdateDatacenterKVPairs(t *testing.T) {
 			URL:    "/v1/txn?dc=dc1",
 			Request: consulapi.TxnOps{
 				{
-					KV: &consulapi.KVTxnOp{Verb: consulapi.KVGet, Key: "test"}, // exists
+					KV: &consulapi.KVTxnOp{Verb: consulapi.KVGet, Key: "test"},
 				},
 				{
-					KV: &consulapi.KVTxnOp{Verb: consulapi.KVGet, Key: "test2"}, // does not exist
+					KV: &consulapi.KVTxnOp{Verb: consulapi.KVGet, Key: "test2"},
 				},
 			},
 			Response: &consulapi.TxnResponse{
 				Results: consulapi.TxnResults{
 					{
-						KV: &consulapi.KVPair{Key: "test", Value: []byte("test")},
+						KV: &consulapi.KVPair{Key: "test", Value: []byte("test"), ModifyIndex: 123},
+					},
+					{
+						KV: &consulapi.KVPair{Key: "test2", Value: []byte("not-equal"), ModifyIndex: 123},
 					},
 				},
 			},
+		},
+		{
+			Method: "PUT",
+			URL:    "/v1/txn?dc=dc1",
+			Request: consulapi.TxnOps{
+				{
+					KV: &consulapi.KVTxnOp{Verb: consulapi.KVGet, Key: "test"},
+				},
+				{
+					KV: &consulapi.KVTxnOp{Verb: consulapi.KVGet, Key: "doesnt-exist"},
+				},
+			},
+			Response: &consulapi.TxnResponse{
+				Errors: consulapi.TxnErrors{
+					{
+						OpIndex: 1,
+						What:    `key "doesnt-exist" doesn't exist`,
+					},
+				},
+			},
+			ResponseCode: http.StatusConflict, // PUT /v1/txn returns a HTTP 409 code on txn failure
 		},
 		{
 			Method: "PUT",
@@ -128,35 +152,78 @@ func TestConsulTxnStoreUpdateDatacenterKVPairs(t *testing.T) {
 				},
 			},
 		},
+		{
+			Method: "PUT",
+			URL:    "/v1/txn?dc=dc1",
+			Request: consulapi.TxnOps{
+				{
+					KV: &consulapi.KVTxnOp{Verb: consulapi.KVSet, Key: "test", Value: []byte("test")},
+				},
+				{
+					KV: &consulapi.KVTxnOp{Verb: consulapi.KVSet, Key: "doesnt-exist", Value: []byte("test")},
+				},
+			},
+			Response: &consulapi.TxnResponse{
+				Results: consulapi.TxnResults{
+					{
+						KV: &consulapi.KVPair{Key: "test", Value: []byte("test")},
+					},
+					{
+						KV: &consulapi.KVPair{Key: "doesnt-exist", Value: []byte("test")},
+					},
+				},
+			},
+		},
 	})
 	defer server.Close()
-	config.Config.ConsulAddress = server.URL
-
-	store := NewConsulTxnStore().(*consulTxnStore)
-	cacheKey := getConsulKVCacheKey(consulTestDefaultDatacenter, "cached")
-	store.kvCache.SetDefault(cacheKey, "cached") // pre-cache the 'cached' key-value
 
 	var wg sync.WaitGroup
-	wg.Add(1)
-	kvPairs := []*consulapi.KVPair{
-		{Key: "cached", Value: []byte("cached")}, // already cached key-value
-		{Key: "test", Value: []byte("test")},     // already correct on consul server
-		{Key: "test2", Value: []byte("test")},    // does not exist on consul server
-	}
-	skipped, existing, written, failed, err := store.updateDatacenterKVPairs(&wg, consulTestDefaultDatacenter, kvPairs)
-	if err != nil {
-		t.Fatalf(".updateDatacenterKVPairs() should not return an error, got: %v", err)
-	}
-	if skipped != 1 || existing != 1 || written != 1 || failed != 0 {
-		t.Fatalf("expected: existing/skipped/written=1 and failed=0, got: skipped=%d, existing=%d, written=%d, failed=%d",
-			skipped, existing, written, failed,
-		)
-	}
+	config.Config.ConsulAddress = server.URL
+	store := NewConsulTxnStore().(*consulTxnStore)
 
-	for _, pair := range kvPairs {
-		cacheKey := getConsulKVCacheKey(consulTestDefaultDatacenter, pair.Key)
-		if cached, found := store.kvCache.Get(cacheKey); !found || cached != string(pair.Value) {
-			t.Fatalf("expected cache key %q to equal %q, got %v", cacheKey, string(pair.Value), cached)
+	t.Run("success-cached", func(t *testing.T) {
+		wg.Add(1)
+		cacheKey := getConsulKVCacheKey(consulTestDefaultDatacenter, "cached")
+		store.kvCache.SetDefault(cacheKey, "cached") // pre-cache the 'cached' key-value
+		defer store.kvCache.Flush()
+
+		kvPairs := []*consulapi.KVPair{
+			{Key: "cached", Value: []byte("cached")}, // already cached key-value
+			{Key: "test", Value: []byte("test")},     // already correct on consul server
+			{Key: "test2", Value: []byte("test")},    // not equal on consul server
 		}
-	}
+		skipped, existing, written, failed, err := store.updateDatacenterKVPairs(&wg, consulTestDefaultDatacenter, kvPairs)
+		if err != nil {
+			t.Fatalf(".updateDatacenterKVPairs() should not return an error, got: %v", err)
+		}
+		if skipped != 1 || existing != 1 || written != 1 || failed != 0 {
+			t.Fatalf("expected: existing/skipped/written=1 and failed=0, got: skipped=%d, existing=%d, written=%d, failed=%d",
+				skipped, existing, written, failed,
+			)
+		}
+
+		for _, pair := range kvPairs {
+			cacheKey := getConsulKVCacheKey(consulTestDefaultDatacenter, pair.Key)
+			if cached, found := store.kvCache.Get(cacheKey); !found || cached != string(pair.Value) {
+				t.Fatalf("expected cache key %q to equal %q, got %v", cacheKey, string(pair.Value), cached)
+			}
+		}
+	})
+
+	t.Run("success-missing-kv", func(t *testing.T) {
+		wg.Add(1)
+		kvPairs := []*consulapi.KVPair{
+			{Key: "test", Value: []byte("test")},         // already correct on consul server
+			{Key: "doesnt-exist", Value: []byte("test")}, // does not exist on consul server
+		}
+		skipped, existing, written, failed, err := store.updateDatacenterKVPairs(&wg, consulTestDefaultDatacenter, kvPairs)
+		if err != nil {
+			t.Fatalf(".updateDatacenterKVPairs() should not return an error, got: %v", err)
+		}
+		if skipped != 0 || existing != 0 || written != 2 || failed != 0 { // confirm all KVs are updated if one does not exist
+			t.Fatalf("expected: existing/skipped/failed=0 and written=2, got: skipped=%d, existing=%d, written=%d, failed=%d",
+				skipped, existing, written, failed,
+			)
+		}
+	})
 }

